dot/sync/worker_pool.go

```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+// punishPeer given a peer.ID we check increase its times punished
+// and apply the punishment time using the base timeout of 5m, so
+// each time a peer is punished its timeout will increase by 5m
+func (s *syncWorkerPool) punishPeer(who peer.ID) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	worker, has := s.workers[who]
+	if !has {
+		return
+	}
+
+	timesPunished := worker.timesPunished + 1
+	punishmentTime := time.Duration(timesPunished) * punishmentBaseTimeout
+	logger.Debugf("⏱️ punishement time for peer %s: %.2fs", who, punishmentTime.Seconds())
+
+	s.workers[who] = &peerSyncWorker{
+		status:         punished,
+		timesPunished:  timesPunished,
+		punishmentTime: time.Now().Add(punishmentTime),
+	}
+}
+
+func (s *syncWorkerPool) ignorePeerAsWorker(who peer.ID) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	delete(s.workers, who)
+	s.ignorePeers[who] = struct{}{}
+}
+
+// totalWorkers only returns available or busy workers
+func (s *syncWorkerPool) totalWorkers() (total uint) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	for _, worker := range s.workers {
+		if worker.status == available {
+			total += 1
+		}
+	}
+
+	return total
+}
+
+// getAvailablePeer returns the very first peer available, if there
+// is no peer avaible then the caller should wait for availablePeerCh
+func (s *syncWorkerPool) getAvailablePeer() peer.ID {
+	for peerID, peerSync := range s.workers {
+		switch peerSync.status {
+		case punished:
+			// if the punishedTime has passed then we can
+			// use it as an available peer
+			if peerSync.punishmentTime.Before(time.Now()) {
+				return peerID
+			}
+		case available:
+			return peerID
+		default:
+		}
+	}
+
+	return peer.ID("")
+}
+
+func (s *syncWorkerPool) getPeerByID(peerID peer.ID) *peerSyncWorker {
+	peerSync, has := s.workers[peerID]
+	if !has {
+		return nil
+	}
+
+	return peerSync
+}
+
+func (s *syncWorkerPool) listenForRequests(stopCh chan struct{}) {
+	defer close(s.doneCh)
+	for {
+		select {
+		case <-stopCh:
+			// wait for ongoing requests to be finished before returning
+			s.wg.Wait()
+			return
+
+		case task := <-s.taskQueue:
+			// whenever a task arrives we try to find an available peer
+			// if the task is directed at some peer then we will wait for
+			// that peer to become available, same happens a normal task
+			// arrives and there is no available peer, then we should wait
+			// for someone to become free and then use it.
+
+			s.mtx.Lock()
+			for {
+				var peerID peer.ID
+				if task.boundTo != nil {
+					peerSync := s.getPeerByID(*task.boundTo)
+					if peerSync != nil && peerSync.status == available {
+						peerID = *task.boundTo
+					}
+				} else {
+					peerID = s.getAvailablePeer()
+				}
+
+				if peerID != peer.ID("") {
+					peerSync := s.workers[peerID]
+					peerSync.status = busy
+					s.workers[peerID] = peerSync
+
+					s.mtx.Unlock()
+
+					s.wg.Add(1)
+					go s.executeRequest(peerID, task)
+					break
+				}
+
+				s.availableCond.Wait()
```

Comment:
I'm confused why we need to use a `sync.Cond` in this case.  In what situation does this line get executed? We will wait on a specific peer to signal on this condition so that that the request can be assigned back to the peer? Maybe some more comments describing the logic here would be best.

Reply:
It makes it easy to handle requests waiting for peers to become available. Basically, when the peers in the pool are all busy then the next task will reach that line so it will wait for the shared `sync.Cond` to signalize when some peer becomes available.

Every worker that finishes handling a task signalizes using `sync.Cond.Signal()` and that is a non-blocking operation.

Reply:
My concern here is that there will be significant back pressure on the `s.taskQueue` in the event that a task is bound to a specific peer, where the `peerSyncWorker` record shows that it's busy.  Essentially no more tasks are pulled off the `taskQueue` channel in the case where line 257 is reached, even though other peers can execute other tasks in parallel.

I've tried to recreate your solution with `sync.Cond` in this [test](https://github.com/ChainSafe/gossamer/blob/6962ada3c3b2c3430f9173ad338a1966bf13b010/dot/sync/worker_pool_test.go#L390).  I've created an alternative solution using channels to represent a queue per peer with a global concurrency limit in this [test](https://github.com/ChainSafe/gossamer/blob/6962ada3c3b2c3430f9173ad338a1966bf13b010/dot/sync/worker_pool_test.go#L425). The second approach with the queue per peer, will execute transactions for all peers in parallel up to the concurrency limit.